### DISCUSSION TITLE
Performance Issues with MMPBSA and MMGBSA Calculations on Multi-Core CPUs

### DISCUSSION THREAD

**Discussion by aprijal99** [Valdes-Tresanco-MS/gmx_MMPBSA#525] [Q&A] [answered]
I am trying to run an MMPBSA calculation on a CPU with 128 cores. However, it is taking a considerable amount of time, processing only about 50 frames in 57 minutes. Additionally, when I run the same calculation using MMGBSA with 8 cores, there is no noticeable difference in performance. For context, my protein has about 800 residues, and the ligand is approximately 400 Da. Could someone help me understand why this might be happening and how I can optimize the calculations for better performance?

**Comment by Valdes-Tresanco-MS** [comment]
gmx_MMPBSA works by splitting the trajectory into many chunks as you assign CPUs. Like all other parallelizations, this type of parallelization has its performance issues. Each process has an associated overhead, so you should be aware of that to get the most out of it. For example, it doesn't make much sense to use 10 CPUs to process 10 frames with GB, since the time it takes to process each frame is less than the time it takes to create the trajectory chunks. Just as you mentioned, with GB, processing 50 frames with 8 CPUs will take very little time. Probably for PB, it is worth increasing the number of CPUs. Please try a longer trajectory and see if it improves according to your criteria. Otherwise, please send me a run schedule with several CPUs up to 16 (that's the number I have available at the moment) and the files you used.

**Reply by aprijal99** [reply]
Thank you for your response.

I'm currently calculating the MMPBSA of a trajectory with 50,000 frames (1 µs), but the process has not completed even after 10 days. It is still at the stage of 'calculating complex contribution.'

Additionally, the 128 cores are part of a single CPU, which I believe is from AMD, and are not distributed across multiple CPUs (not parallel).

You can find the files I used in the following link:
https://drive.google.com/drive/folders/1SX12IxMxVw5USqVpUmyrsbfyTNTvL2LQ

**Reply by Valdes-Tresanco-MS** [reply]
Please, attach the gmx_MMPBSA.log file to check your config

**Reply by aprijal99** [reply]
Here is the gmx_MMPBSA.log file, I also uploaded it to the above link

[gmx_MMPBSA.log](https://github.com/user-attachments/files/16401605/gmx_MMPBSA.log)

**Reply by Valdes-Tresanco-MS** [reply, marked as answer]
Here you aren't using parallelization. To run gmx_MMPBSA in parallel you must run it with `MPI`. In your  case should be:
```bash
mpirun -np 128 gmx_MMPBSA -O -i mmpbsa.in -cs prot_ligand.tpr -ci index.ndx -cg 1 13 -ct md_final_whole_nojump_mol.xtc -cp topol.top -o FINAL_RESULTS_MMPBSA.dat -eo FINAL_RESULTS_MMPBSA.csv
```

**Reply by aprijal99** [reply]
Okay, I'll try using the mpirun command and will update you on the performance

Initially, I thought that mpirun was only necessary for running MMPBSA calculations on a multi-node computer cluster

**Reply by aprijal99** [reply]
Thank you for your help. I've noticed a significant increase in the speed of the MMGBSA calculation, now processing about 5,000 frames in 30 minutes

I haven't tried the MMPBSA calculation yet, but I hope it will also show a reduction in processing time
